contracts/protocol/v1/proxies/P1LiquidatorProxy.sol

```diff
@@ -0,0 +1,285 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.5.16;
+pragma experimental ABIEncoderV2;
+
+import { Ownable } from "@openzeppelin/contracts/ownership/Ownable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { BaseMath } from "../../lib/BaseMath.sol";
+import { SignedMath } from "../../lib/SignedMath.sol";
+import { I_PerpetualV1 } from "../intf/I_PerpetualV1.sol";
+import { P1BalanceMath } from "../lib/P1BalanceMath.sol";
+import { P1Types } from "../lib/P1Types.sol";
+
+
+/**
+ * @title P1LiquidatorProxy
+ * @author dYdX
+ *
+ * @notice Proxy contract for liquidating accounts. A fixed percentage of each liquidation is
+ * directed to the insurance fund.
+ */
+contract P1LiquidatorProxy is
+    Ownable
+{
+    using BaseMath for uint256;
+    using SignedMath for SignedMath.Int;
+    using P1BalanceMath for P1Types.Balance;
+
+    // ============ Events ============
+
+    event LogLiquidatorProxyUsed(
+        address indexed liquidator,
+        address indexed liquidatee,
+        bool isBuy,
+        uint256 liquidationAmount,
+        uint256 feeAmount
+    );
+
+    event LogInsuranceFundSet(
+        address insuranceFund
+    );
+
+    // ============ Immutable Storage ============
+
+    // Address of the perpetual contract.
+    address public _PERPETUAL_V1_;
+
+    // Address of the P1Liquidation contract.
+    address public _LIQUIDATION_;
+
+    // Percentage of fees that go to the insurance fund as a fixed-point number with 18 decimals.
+    uint256 public _FEE_PERCENTAGE_;
+
+    // ============ Mutable Storage ============
+
+    // Address of the insurance fund.
+    address public _INSURANCE_FUND_;
+
+    // ============ Constructor ============
+
+    constructor (
+        address perpetualV1,
+        address liquidator,
+        address insuranceFund,
+        uint256 feePercentage
+    )
+        public
+    {
+        _PERPETUAL_V1_ = perpetualV1;
+        _LIQUIDATION_ = liquidator;
+        _INSURANCE_FUND_ = insuranceFund;
+        _FEE_PERCENTAGE_ = feePercentage;
+
+        emit LogInsuranceFundSet(insuranceFund);
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Sets the maximum allowance on the perpetual contract. Must be called once.
+     * @dev Cannot be run in the constructor due to technical restrictions in Solidity.
+     */
+    function setAllowance()
+        external
+    {
+        address perpetual = _PERPETUAL_V1_;
+        address tokenContract = I_PerpetualV1(perpetual).getTokenContract();
+        SafeERC20.safeApprove(
+            IERC20(tokenContract),
+            perpetual,
+            uint256(-1)
```

Comment:
Do you know if USDC has the thing where approving the maximum int sets unlimited allowance? Otherwise it's possible this could run out and have to be called again

Reply:
It's not really possible as it's such a high number, but will change this a bit due to some weirdness with `SafeERC20.safeApprove()`

Reply:
I don't think so. The way `SafeERC20` works, `safeApprove` is actually likely to fail if called again. But I didn't think this was an issue since it's 10^71 dollars or 10^59 DAI.